crypto/data_formats.go

```diff
@@ -15,41 +15,68 @@
 package crypto
 
 import (
-	"encoding/base64"
 	"fmt"
-	"strconv"
 
-	"github.com/benlaurie/objecthash/go/objecthash"
+	"github.com/google/certificate-transparency-go/tls"
+
 	"github.com/google/trillian"
 )
 
 // This file contains struct specific mappings and data structures.
-// TODO(gdbelvin): remove data-structure specific operations.
-
-// Constants used as map keys when building input for ObjectHash. They must not be changed
-// as this will change the output of hashRoot()
-const (
-	mapKeyRootHash       string = "RootHash"
-	mapKeyTimestampNanos string = "TimestampNanos"
-	mapKeyTreeSize       string = "TreeSize"
-)
 
-// HashLogRoot hashes SignedLogRoot objects using ObjectHash with
-// "RootHash", "TimestampNanos", and "TreeSize", used as keys in
-// a map.
-func HashLogRoot(root trillian.SignedLogRoot) ([]byte, error) {
-	// Pull out the fields we want to hash.
-	// Caution: use string format for int64 values as they can overflow when
-	// JSON encoded otherwise (it uses floats). We want to be sure that people
-	// using JSON to verify hashes can build the exact same input to ObjectHash.
-	rootMap := map[string]interface{}{
-		mapKeyRootHash:       base64.StdEncoding.EncodeToString(root.RootHash),
-		mapKeyTimestampNanos: strconv.FormatInt(root.TimestampNanos, 10),
-		mapKeyTreeSize:       strconv.FormatInt(root.TreeSize, 10)}
-
-	hash, err := objecthash.ObjectHash(rootMap)
-	if err != nil {
-		return nil, fmt.Errorf("ObjectHash(%#v): %v", rootMap, err)
+// SignedLogRootV1 contains the fields verified by SignedLogRootV1
+type SignedLogRootV1 struct {
+	DataFormatVersion uint32
+	RootHash          []byte `tls:"minlen:0,maxlen:128"`
+	TimestampNanos    uint64
+	TreeSize          uint64
+	LogID             uint64
+}
+
+// SignedMapRootV1 contains the fields verified by SignedMapRootV1
+type SignedMapRootV1 struct {
+	DataFormatVersion uint32
```

Comment:
If you chose to slay the 32 bit dragon on page 1, attack the 2nd dragon here and then turn to page 4 to continue your quest.